packages/@aws-cdk/aws-cloudformation/test/test.pipeline-actions.ts

```diff
@@ -0,0 +1,144 @@
+      const stack = new cdk.Stack();
+      const statements = new Array<PolicyStatementJson>();
+      const pipelineRole = {
+        addToPolicy: statement => { statements.push(statement.toJson()); }
+      } as iam.Role;
+      const actions = new Array<cpapi.Action>();
+      const stage = {
+        _attachAction: act => { actions.push(act); },
+        pipelineRole
+      } as cpapi.IStage;
+      new cloudformation.PipelineExecuteChangeSetAction(stack, 'Action', {
+        stage,
+        changeSetName: 'MyChangeSet',
+        stackName: 'MyStack',
+      });
+
+      const stackArn = cdk.ArnUtils.fromComponents({
+        service: 'cloudformation',
+        resource: 'stack',
+        resourceName: 'MyStack/*'
+      });
+      test.ok(_grantsPermission(statements, 'cloudformation:ExecuteChangeSet', stackArn, {
+                StringEquals: { 'cloudformation:ChangeSetName': 'MyChangeSet' }
+              }),
+              'The pipelineRole was given permissions to execute the desired ChangeSet');
+
+      test.ok(_hasAction(actions, 'AWS', 'CloudFormation', 'Deploy', {
+        ActionMode: 'CHANGE_SET_EXECUTE',
+        StackName: 'MyStack',
+        ChangeSetName: 'MyChangeSet'
+      }));
+
+      test.done();
+    }
+  }
+});
+
+interface PolicyStatementJson {
+  Effect: 'Allow' | 'Deny';
+  Action: string | string[];
+  Resource: string | string[];
+  Condition: any;
+}
+
+function _hasAction(actions: cpapi.Action[], owner: string, provider: string, category: string, configuration?: { [key: string]: any}) {
+  for (const action of actions) {
+    if (action.owner !== owner) { continue; }
+    if (action.provider !== provider) { continue; }
+    if (action.category !== category) { continue; }
+    if (configuration && !action.configuration) { continue; }
+    if (configuration) {
+      for (const key of Object.keys(configuration)) {
+        if (!util.isDeepStrictEqual(cdk.resolve(action.configuration[key]), cdk.resolve(configuration[key]))) {
+          continue;
+        }
+      }
+    }
+    return true;
+  }
+  return false;
+}
+
+function _grantsPermission(statements: PolicyStatementJson[], action: string, resource: string, conditions?: any) {
+  for (const statement of statements.filter(s => s.Effect === 'Allow')) {
+    if (!_isOrContains(statement.Action, action)) { continue; }
+    if (!_isOrContains(statement.Resource, resource)) { continue; }
+    if (conditions && !_isOrContains(statement.Condition, conditions)) { continue; }
+    return true;
+  }
+  return false;
+}
+
+function _isOrContains(entity: string | string[], value: string): boolean {
+  const resolvedValue = cdk.resolve(value);
+  const resolvedEntity = cdk.resolve(entity);
+  if (util.isDeepStrictEqual(resolvedEntity, resolvedValue)) { return true; }
+  if (!Array.isArray(resolvedEntity)) { return false; }
+  for (const tested of entity) {
+    if (util.isDeepStrictEqual(tested, resolvedValue)) { return true; }
+  }
+  return false;
```

Comment:
Yeah...

Integ test instead? 🙃

Reply:
My perception here (prove me wrong if you can :D) is that those are super brittle, because they involve almost as many, if not more, parts that are not part of the tested module than parts that I want to test. It is also not possible for me to integ-test from `@aws-cdk/aws-cloudformation` without incurring dependency cycles, and moving the test to somewhere else (`aws-cdk/aws-codepipeline`) feels wrong & would mis-represent the test coverage if the CFN L2, which I think is a problem.

Reply:
Do you think it's possible to substitute some of these helper functions with our `expect` `haveResource` helpers from `cdk-assert`? I'm worried that the failures these produce will be pretty much impossible to diagnose (for example, when you return `false` from `_hasAction`).

Reply:
That's what we have stack traces for, but I hear you. The problem will be that `expect` / `haveResource` operate on *synthesized* stacks, and that I am precisely unable to synthesize a stack here because it requires me to pull in many unneeded things. Instead, I propose to make assertion-style helpers that will actually format the *actual* in the message, so it is more actionable.